### PR TITLE
Refactor On Eldritch Surge

### PR DESCRIPTION
**Pull request by HIEROT**
An intricate and brand new subclass! Much more interesting to play than just EB like before.
Fix some bugs and Add some new interfaces.
Major change to original code:
Move Spiritual Shield blocking and Flash Of Genius save bonus to thier separate class in favor of new interfaces.
Make IDefinitionCustomCode applicable to invocation's granted feature to make them able to do things when added and removed. Meanwhile the relevent codes in Party Editor and Level Up Context are modified.
New interfaces:
IInvocationToggled, IMeOrAllySaveFailPossible (Flash Of Genius), IModifyAC, ISpellCast, IAttackBeforeHitPossibleOnMeOrAlly (Spiritual Shield)
New infrastracture:
RulesetConditionCustom
Original custom condition used in Blast Overload is created with new instance when added to character, however, it will return to a pool of RulesetCondition when removed and possibly add to other character in the future, which will eventually mess up your Save (though it will not crash). Newly implemented RulesetConditionCustom has its own pool and you can use it safely and conveniently just as original RulesetCondition, as long as you follow my practice (which is also noted down in the comments).

TODO:
Add texts ans sprite resources. (Done)

### CODE REVIEW

**Comment by HIEROT** [comment]
Please merge others' PR first cause I renamed some methods for clarity.

File: SolastaUnfinishedBusiness/Patches/GameLocationBattleManagerPatcher.cs [hunk 1]
```diff
@@ -458,13 +458,24 @@ int attackRoll
...
+                var rulesetDefender = defender.RulesetCharacter;
+
+                foreach (var featureOwner in Gui.Battle.GetOpposingContenders(attacker.Side)
+                             .Where(u => u.RulesetCharacter is { IsDeadOrDyingOrUnconscious: false })
+                             .ToList())
 
-                while (blockEvents.MoveNext())
                 {
-                    yield return blockEvents.Current;
+                    foreach (var extraEvents in featureOwner.RulesetCharacter
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Patches/GameLocationBattleManagerPatcher.cs:470]: Here is the change made to this file, others are reformat

File: SolastaUnfinishedBusiness/Patches/GameLocationBattleManagerPatcher.cs [hunk 2]
```diff
@@ -1044,81 +1055,28 @@ bool hasBorrowedLuck
...
-                {
-                    continue;
-                }
-
-                var power = feature.Power;
-
-                if (!helper.CanUsePower(power))
+                foreach (var extraEvents in featureOwner.RulesetCharacter
+                        .GetSubFeaturesByType<IMeOrAllySaveFailPossible>()
+                        .Select(x => x.OnMeOrAllySaveFailPossible(__instance, action, attacker, defender, featureOwner, saveModifier, hasHitVisual,
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Patches/GameLocationBattleManagerPatcher.cs:1069]: Here is the second change.
IMeOrAllySaveFailPossible

File: SolastaUnfinishedBusiness/Patches/GameLocationBattleManagerPatcher.cs [hunk 3]
```diff
@@ -1355,4 +1313,37 @@ public static class HandleCharacterPhysicalAttackFinished_Patch
...
+    public static class HandleSpellCast_Patch
+    {
+        [UsedImplicitly]
+        public static IEnumerator Postfix(
+            IEnumerator values,
+            GameLocationCharacter caster,
+            CharacterActionCastSpell castAction,
+            RulesetEffectSpell selectEffectSpell,
+            RulesetSpellRepertoire selectedRepertoire,
+            SpellDefinition selectedSpellDefinition)
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Patches/GameLocationBattleManagerPatcher.cs:1329]: Here is the third,
ISpellCast

File: SolastaUnfinishedBusiness/Patches/LocalCommandManagerPatcher.cs [hunk 4]
```diff
@@ -37,4 +38,33 @@ private static void Check(GameLocationCharacter character, ActionDefinitions.Act
...
+    [UsedImplicitly]
+    public static class TogglePermanentInvocation_Patch
+    {
+        [NotNull]
+        [UsedImplicitly]
+
+        public static bool Prefix(GameLocationCharacter character, RulesetInvocation invocation)
+        {
+            if (character != null)
+            {
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Patches/LocalCommandManagerPatcher.cs:53]: IInvocationToggled

File: SolastaUnfinishedBusiness/Patches/RulesetActorPatcher.cs [hunk 5]
```diff
@@ -35,14 +37,27 @@ public static class AddConditionOfCategory_Patch
...
             {
                 return;
             }
 
+            // Enable RulesetConditionCustom
+            var replaceWithRulesetConditionCustom = newCondition.conditionDefinition.GetFirstSubFeatureOfType<IBindToRulesetConditionCustom>();
+            if (replaceWithRulesetConditionCustom != null)
+            {
+                var originalCondtion = newCondition;
+                // The original condition is yet to register, however it is got from its pool, so we should return it
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Patches/RulesetActorPatcher.cs:53]: IBindToRulesetConditionCustom

File: SolastaUnfinishedBusiness/Patches/RulesetActorPatcher.cs [hunk 6]
```diff
@@ -799,4 +814,76 @@ public static IEnumerable<CodeInstruction> Transpiler([NotNull] IEnumerable<Code
                 new CodeInstruction(OpCodes.Callvirt, replacingMethod));
         }
     }
+
+    //PATCH: allow ISpellAffinityProvider to be validated with IRemoveSpellOrSpellLevelImmunity
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Patches/RulesetActorPatcher.cs:818]: IRemoveSpellOrSpellLevelImmunity

File: SolastaUnfinishedBusiness/Patches/RulesetCharacterHeroPatcher.cs [hunk 7]
```diff
@@ -111,6 +111,33 @@ public static IEnumerable<CodeInstruction> Transpiler([NotNull] IEnumerable<Code
 
             return codes;
         }
+        [UsedImplicitly]
+        public static void Postfix(
+            RulesetCharacterHero __instance,
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Patches/RulesetCharacterHeroPatcher.cs:116]: IModifyAC

File: SolastaUnfinishedBusiness/Patches/RulesetCharacterMonsterPatcher.cs [hunk 8]
```diff
@@ -90,6 +90,33 @@ public static IEnumerable<CodeInstruction> Transpiler([NotNull] IEnumerable<Code
                 new CodeInstruction(OpCodes.Ldarg_0),
                 new CodeInstruction(OpCodes.Call, unstack));
         }
+        [UsedImplicitly]
+        public static void Postfix(
+            RulesetCharacterMonster __instance,
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Patches/RulesetCharacterMonsterPatcher.cs:95]: IModifyAC

File: SolastaUnfinishedBusiness/Subclasses/MartialRoyalKnight.cs [hunk 9]
```diff
@@ -276,22 +278,66 @@ internal InspiringProtection(FeatureDefinitionPower power, string reactionName)
...
+            GameLocationCharacter helper)
         {
             var text = defender == helper
                 ? "Reaction/&SpendPowerRoyalKnightInspiringProtectionDescriptionSelf"
                 : "Reaction/&SpendPowerRoyalKnightInspiringProtectionDescriptionAlly";
 
             return Gui.Format(text, defender.Name, attacker.Name, action.FormatTitle());
         }
+
+        public IEnumerator OnMeOrAllySaveFailPossible(GameLocationBattleManager battleManager, CharacterAction action, GameLocationCharacter attacker, GameLocationCharacter defender, GameLocationCharacter featureOwner, ActionModifier saveModifier, bool hasHitVisual, bool hasBorrowedLuck)
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Subclasses/MartialRoyalKnight.cs:294]: Inspiring Protection logic moved here

File: SolastaUnfinishedBusiness/Subclasses/OathOfAltruism.cs [hunk 10]
```diff
@@ -184,4 +187,106 @@ public IEnumerator OnUsePowerFinishedByMe(CharacterActionUsePower action, Featur
             }
         }
     }
+
+    private class SpiritualShieldingBlockAttack: IAttackBeforeHitPossibleOnMeOrAlly
+    {
+        public IEnumerator OnAttackBeforeHitPossibleOnMeOrAlly(GameLocationBattleManager battleManager, GameLocationCharacter featureOwner, GameLocationCharacter attacker, GameLocationCharacter defender, RulesetAttackMode attackMode, RulesetEffect rulesetEffect, ActionModifier attackModifier, int attackRoll)
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Subclasses/OathOfAltruism.cs:193]: Spiritual Shield logic moved here

File: SolastaUnfinishedBusiness/Classes/InventorClass.cs [hunk 11]
```diff
@@ -965,21 +963,72 @@ private static int GetBonus(RulesetEntity helper)
...
     {
         var text = defender == helper
             ? "Reaction/&SpendPowerInventorFlashOfGeniusReactDescriptionSelfFormat"
             : "Reaction/&SpendPowerInventorFlashOfGeniusReactAllyDescriptionAllyFormat";
 
         return Gui.Format(text, defender.Name, attacker.Name, action.FormatTitle());
     }
+
+    public IEnumerator OnMeOrAllySaveFailPossible(GameLocationBattleManager battleManager,
+        CharacterAction action,
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/Classes/InventorClass.cs:980]: Flash of Genius logic move here

File: SolastaUnfinishedBusiness/CustomBuilders/EldritchVersatility.cs [hunk 12]
```diff
@@ -0,0 +1,1269 @@
...
+using static ActionDefinitions;
+using SolastaUnfinishedBusiness.Api.Infrastructure;
+using SolastaUnfinishedBusiness.Models;
+using SolastaUnfinishedBusiness.Api.LanguageExtensions;
+using TA;
+using static FeatureDefinitionAttributeModifier;
+
+namespace SolastaUnfinishedBusiness.CustomBuilders;
+static class EldritchVersatility
+{
```
Comment (HIEROT) [inline SolastaUnfinishedBusiness/CustomBuilders/EldritchVersatility.cs:28]: Main file for new logic

**Review by ThyWoof — APPROVED** [review]

**Pull request merged**